### DISCUSSION TITLE
I struggle to grasp transactions, could someone explain the concept to me?

### DISCUSSION THREAD

**Discussion by Weyaaron** [sqlalchemy/sqlalchemy#7358] [Usage Questions]
Dear Community,

ever since using sqlalchemy, I struggled with transactions. Since this is such a fundamental concept, this causes a lot of stress. But even after some conscious efforts to learn this I still make the same errors most of the time, and when it does work, I cant quite wrap my head around whats different. I will add the example that tripped me over this time and add the points I don't get. 

My goal is to delete all of the articles whose publisher_id is 9. Ignoring the transactions for a moment, this is straightforward:
```
backend_url: str = construct_postgres_url(**DatabaseConfig.backend_admin)
backend_engine = create_engine(backend_url, echo=False, future=True)

BackendSession = sessionmaker(bind=backend_engine)
with BackendSession() as backend_session:

    bild_articles: List[Article] = backend_session.query(Article).filter_by(publisher_id=9).all()
    for article_el in bild_articles:
            backend_session.delete(article_el)
```
However, the code above does not do anything, since the changes are not commited. (As far as I can tell)
One could use one big transaction by doing this:
```
with BackendSession.begin() as backend_session:
...
```
However, this solution has the disadvantage that all of the articles are deleted in one transaction. I would prefer a solution that deletes each article on its own. And this is where I run into trouble:
```
with BackendSession.begin() as backend_session:

    bild_articles: List[Article] = backend_session.query(Article).filter_by(publisher_id=9).all()
    for article_el in bild_articles:
        with backend_session.begin():
            backend_session.delete(article_el)
```
This code crashes with a "InvalidRequestError", since a transaction has already started. Which I sort of get, since the first begin starts a transaction.
```
with BackendSession() as backend_session:

    bild_articles: List[Article] = backend_session.query(Article).filter_by(publisher_id=9).all()
    for article_el in bild_articles:
        with backend_session.begin():
            backend_session.delete(article_el)
```
This crashes as well, since the Query-Statement starts a transaction and the first transaction that deletes an Article is invalid, since a transaction has been started. But I dont quite get how or why this query statement starts a transaction. 
```
with BackendSession() as backend_session:
    with backend_session.begin():
        bild_articles: List[Article] = backend_session.query(Article).filter_by(publisher_id=9).all()
    for article_el in bild_articles:
        with backend_session.begin():
            backend_session.delete(article_el)
```
This code finally does the job. But I dont quite get whats the difference. Obviously, the context-manager in the second line wraps the transactions such that there is one active transaction at all times but no more. 
However, I dont understand why this is necessary? 
I hope my post is detailed enough to express my confusion, any help is appreciated.
Have a nice day,
Aaron

**Comment by gordthompson** [comment]
As mentioned in the [tutorial](https://docs.sqlalchemy.org/en/14/tutorial/dbapi_transactions.html#executing-with-an-orm-session) you can use "commit as you go" to delete the articles one by one:

```python
with Session(engine, future=True) as backend_session:
    bild_articles: List[Article] = backend_session.query(Article).filter_by(publisher_id=9).all()
    for article_el in bild_articles:
        backend_session.delete(article_el)
        backend_session.commit()
```

which generates the following SQL (as shown if we use `echo=True` in the `create_engine()` call):

```
2021-11-23 09:56:56,009 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-11-23 09:56:56,012 INFO sqlalchemy.engine.Engine SELECT article.id AS article_id, article.publisher_id AS article_publisher_id 
FROM article 
WHERE article.publisher_id = ?
2021-11-23 09:56:56,012 INFO sqlalchemy.engine.Engine [generated in 0.00028s] (9,)
2021-11-23 09:56:56,014 INFO sqlalchemy.engine.Engine DELETE FROM article WHERE article.id = ?
2021-11-23 09:56:56,015 INFO sqlalchemy.engine.Engine [generated in 0.00051s] (1,)
2021-11-23 09:56:56,015 INFO sqlalchemy.engine.Engine COMMIT
2021-11-23 09:56:56,016 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2021-11-23 09:56:56,018 INFO sqlalchemy.engine.Engine SELECT article.id AS article_id, article.publisher_id AS article_publisher_id 
FROM article 
WHERE article.id = ?
2021-11-23 09:56:56,018 INFO sqlalchemy.engine.Engine [generated in 0.00028s] (3,)
2021-11-23 09:56:56,019 INFO sqlalchemy.engine.Engine DELETE FROM article WHERE article.id = ?
2021-11-23 09:56:56,019 INFO sqlalchemy.engine.Engine [cached since 0.004675s ago] (3,)
2021-11-23 09:56:56,019 INFO sqlalchemy.engine.Engine COMMIT
```

**Reply by CaselIT** [reply]
Also note that you can delete all articles without iterating over them by using
`backend_session.query(Article).filter_by(publisher_id=9).delete()`.

This will issue a single delete to the db with the proper where
